### PR TITLE
Improvements to Ruby SDK and Rails demo app

### PR DESCRIPTION
**Pull request by ccschmitz**
## Summary

Creates a new Rails demo app with a page that has buttons to trigger the creation of traces, errors, and logs. Also makes some updates to the Ruby SDK:

* Does not error if methods are called and SDK is not initialized (helpful to code that is run in isolation during tests)
* Adds a `start_span` method that allows you to create a new span with just a name and attributes (the old `trace` method required a session and request ID to be passed, which didn't make sense outside of the controller action)
* Fixes a bug with the logger so its API is compatible with the silent logger (this was causing an issue in the old demo app loading assets)
* Adds some convenience methods to improve the ergonomics of using the SDK, allowing you to call methods directly on the `Highlight` module (e.g. call `Highlight.start_trace` rather than `Highlight::H.instance.start_trace`)
* Enables all available auto instrumentations.

Closes SUP-68

## How did you test this change?

testing using local ruby SDK and local firstload SDK
https://www.loom.com/share/572698c9f81f49579ab3171fe35003d5

image

Screenshot 2024-08-13 at 14 25 43

## Are there any deployment considerations?

new ruby sdk version

## Does this work require review from our design team?

no

### CODE REVIEW

**Comment by ccschmitz** [comment]
@Vadman97 I just cleaned up a few things in this PR in preparation for a handoff to you. I also tossed in 6bc543330e04dd0a44da854b607bc85e7f7eea88 which includes a `<meta>` tag attempting to handoff the server trace to the client. Hadn't successfully connected any client and server traces that way yet, but I think that's how it's supposed to work 😸

**Comment by ccschmitz** [comment]
@Vadman97 was also wondering if you think we should move the Rails app from the SDK directory to the root-level `e2e` directory?

File: sdk/client/src/otel/index.ts [hunk 1]
```diff
@@ -285,7 +285,7 @@ const getSpanName = (
 	let parsedBody
 	const urlObject = new URL(url)
 	const pathname = urlObject.pathname
-	let spanName = `${method} - ${pathname}`
+	let spanName = `${method.toUpperCase()} - ${pathname}`
```
Comment (Vadman97) [inline sdk/client/src/otel/index.ts:288]: @ccschmitz this change is unrelated, right?

Comment (ccschmitz) [reply sdk/client/src/otel/index.ts:288]: Oh yeah, sorry... Can back this change out probably

File: docker/collector.yml
Comment (Vadman97) [inline docker/collector.yml:1]: @ccschmitz i'm noticing a difference between this file and `deploy/otel-collector.yaml`, our prod version. notably in `allowed_headers` - should i update the `deploy` one?

image

File: sdk/client/src/otel/index.ts (hunk 1 above)
Comment (Vadman97) [reply sdk/client/src/otel/index.ts:288]: all good, can also have this here but i'll add a changelog to release the client

File: docker/collector.yml
Comment (ccschmitz) [reply docker/collector.yml:1]: Ah, yes, I think we'll need it in the deploy one as well. Good catch!

File: .github/workflows/ruby.yml [hunk 2]
```diff
@@ -14,6 +14,9 @@ jobs:
     build:
         name: Ruby SDK
         runs-on: ubuntu-latest
+        strategy:
+            matrix:
+                ruby: ['3.0', '3.3']
```
Comment (ccschmitz) [inline .github/workflows/ruby.yml:19]: Nice moves 🕺

File: e2e/ruby/rails/demo/Gemfile [hunk 3]
```diff
@@ -68,7 +68,6 @@ group :test do
   # Use system testing [https://guides.rubyonrails.org/testing.html#system-testing]
   gem "capybara"
   gem "selenium-webdriver"
-  gem "webdrivers"
 end
 
-gem "highlight_io", path: "../../../highlight"
+gem "highlight_io", path: "../../../../sdk/highlight-ruby/highlight"
```
Comment (ccschmitz) [inline e2e/ruby/rails/demo/Gemfile:73]: Maybe we should not change this path?

File: sdk/client/src/utils/storage.ts [hunk 4]
```diff
@@ -34,15 +61,21 @@ export const setStorageMode = (m: Mode) => {
...
+export const setCookieWriteEnabled = (enabled: boolean) => {
+	cookieWriteEnabled = enabled
+}
+
 export const getItem = (key: string) => {
 	return getPersistentStorage().getItem(key)
 }
 
 export const setItem = (key: string, value: string) => {
+	cookieStorage.setItem(key, value)
```
Comment (ccschmitz) [inline sdk/client/src/utils/storage.ts:73]: Sorry if I'm missing something here, but why write everything to cookies too?

File: e2e/ruby/rails/demo/Gemfile (hunk 3 above)
Comment (Vadman97) [reply e2e/ruby/rails/demo/Gemfile:73]: this was from moving the e2e ruby app to the `e2e` directory; the sdk didn't move. does that make sense?

Comment (ccschmitz) [reply e2e/ruby/rails/demo/Gemfile:73]: Oh, I see - sorry!

File: sdk/client/src/utils/storage.ts (hunk 4 above)
Comment (ccschmitz) [reply sdk/client/src/utils/storage.ts:73]: Ah, right - makes sense!

**Review by Vadman97 — APPROVED** [review]
💪
